server/src/main/java/org/elasticsearch/cluster/metadata/IndexNameExpressionResolver.java

```diff
@@ -240,11 +241,25 @@ public IndexAbstraction resolveWriteIndexAbstraction(ClusterState state, DocWrit
         }
     }
 
-    private static Collection<String> resolveExpressions(Context context, String... expressions) {
-        return WildcardExpressionResolver.resolve(
-            context,
-            DateMathExpressionResolver.resolve(context, expressions == null ? List.of() : List.of(expressions))
-        );
+    protected static Collection<String> resolveExpressions(Context context, String... expressions) {
+        if (context.getOptions().expandWildcardExpressions() == false) {
+            if (expressions == null || expressions.length == 0 || expressions.length == 1 && Metadata.ALL.equals(expressions[0])) {
+                return List.of();
+            } else {
+                return ExplicitResourceNameFilter.filter(context, DateMathExpressionResolver.resolve(context, List.of(expressions)));
+            }
+        } else {
+            if (expressions == null
+                || expressions.length == 0
+                || expressions.length == 1 && (Metadata.ALL.equals(expressions[0]) || Regex.isMatchAllPattern(expressions[0]))) {
+                return WildcardExpressionResolver.resolveAll(context);
+            } else {
+                return WildcardExpressionResolver.resolve(
+                    context,
+                    ExplicitResourceNameFilter.filter(context, DateMathExpressionResolver.resolve(context, List.of(expressions)))
```

Comment:
This is now the code path for regular expressions.
Three stages:
* deal with datemaths
* deal with explicit (non-wildcard, non-exclusion) names
* deal with wildcards